README.md
## Surveyweights

Apply Census weighting to survey data.

### Example Usage

```Python
from surveyweights import run_weighting_scheme, run_weighting_iteration

# Define what to weigh on
weigh_on = ['age', 'education', 'gender', 'income', 'race', 'urban_rural', 'vote2016']

# Run weighting
output = run_weighting_scheme(survey_data, iters=25, weigh_on=weigh_on)

# Get data back with weight column
survey_data = output['final_df']

# See balance of weights 
run_weighting_iteration(survey_data, weigh_on=weigh_on)

# Look at unweighted outcome data
print(survey_data['outcome'].value_counts(normalize=True) * 100)

# Look at weighted outcome data
print(survey_data['outcome'].value_counts(normalize=True) * survey_data.groupby('outcome')['weight'].mean() * 100)
```


### Debugging

**Help! The percentages don't sum to 100%!**

If you subset the dataset, you subset the weights too and they will no longer work for the subsetted dataset. To fix this, use `nomalize_weights`:

```Python
# Subset df
subset_df = df[df[var] == subset]

# Look at weighted data (will be wrong and will not sum to 100%!)
print(subset_df[var].value_counts(normalize=True) * subset_df.groupby(var)['weight'].mean() * 100)

# Normalize weights
df['weight'] = nomalize_weights(df['weight'])

# Look at weighted data (it is now fixed and still representative!)
print(subset_df[var].value_counts(normalize=True) * subset_df.groupby(var)['weight'].mean() * 100)
```

~

**Help! The percentages still don't sum to 100% and I used `normalize_weights`!**

Another issue might be missing values. Try removing those.

```Python
df = df.dropna() # Remove NAs
df['weight'] = nomalize_weights(df['weight']) # Normalize weights

# Look at weighted data (it is now fixed and still representative!)
print(subset_df[var].value_counts(normalize=True) * subset_df.groupby(var)['weight'].mean() * 100)
```

Note that you may prefer to drop NAs just for particular columns of interest, or you may prefer to impute NAs with a particular value.


~

**Help! Re-running changes my results!**

The results should be deterministic, so re-running should not affect results. However, the weights still might be unstable and running the same weights in a different order could affect results. To fix this, try increasing the number of iterations and turning off early termination. Also, keep in mind that fluctuations of ~0.1 percentage point could be very normal - potentially a larger fluctuation for very small sample sizes.


### Installation

`pip3 install surveyweights`



CHANGES.md
## v0.6

* Adds `vote_brexit` (2016 Brexit vote), `income`, `education`, and `region` to the list of weights for the UK census.

## v0.5

* Adds `vote2020` to the list of weights for the US national census.

## v0.4

* Fixed a bug where NAs in the dataset would make the package crash. Now it raises an error message.

## v0.3

* `get_census` can now be used to view the census data.
* Fixed a bug where `max_weight` and `min_weight` would not be populated if `verbose=False`

## v0.2

* California census data can be accessed via `census='US_CA'`
* Texas census data can be accessed via `census='US_TX'`
* Custom census cwn be defined via passed dict (`census=custom_dict`)
* Early termination can be disabled by passing `early_terminate=False` to `run_weighting_scheme`


surveyweights/census.py
CENSUS = {'age': {'18-24': 0.1304,
                  '25-44': 0.3505,
                  '45-64': 0.3478,
                  '65+': 0.1713},
          'education': {'Completed graduate school': 0.1204,
                        'Graduated from college': 0.2128,
                        'Some college, no degree': 0.2777,
                        'Graduated from high school': 0.2832,
                        'Less than high school': 0.1060},
          'income': {'Under $15,000': 0.1020,
                     'Between $15,000 and $49,999': 0.2970,
                     'Between $50,000 and $74,999': 0.1720,
                     'Between $75,000 and $99,999': 0.1250,
                     'Between $100,000 and $150,000': 0.1490,
                     'Over $150,000': 0.1550},
          'race': {'White or Caucasian': 0.6340,
                   'Asian or Asian American': 0.0590,
                   'Black or African American': 0.1340,
                   'Hispanic or Latino': 0.1530,
                   'Other': 0.02},
          'urban_rural': {'Suburban': 0.5500,
                          'Urban': 0.3100,
                          'Rural': 0.1400},
          'region': {'Midwest': 0.2149,
                     'Mountains': 0.0539,
                     'Northeast': 0.1604,
                     'Pacific': 0.1602,
                     'South': 0.2180,
                     'Southwest': 0.0531,
                     'Southeast': 0.1398},
          'left_right': {'Liberal': 0.35,  # Sienna College / NYT poll <https://int.nyt.com/data/documenttools/nyt-siena-poll-methodology-june-2020/f6f533b4d07f4cbe/full.pdf>
                         'Moderate': 0.26,
                         'Conservative': 0.30}}




surveyweights/weighting.py
import numpy as np
import pandas as pd

from surveyweights.census.us_census import US_CENSUS, US_CA_CENSUS, US_TX_CENSUS
from surveyweights.census.uk_census import UK_CENSUS


def get_census(census='US'):
    if isinstance(census, dict):
        return census
    if census == 'US':
        return US_CENSUS
    if census == 'US_CA':
        return US_CA_CENSUS
    if census == 'US_TX':
        return US_TX_CENSUS
    elif census == 'UK':
        return UK_CENSUS
    else:
        raise ValueError('{} census not found'.format(census))


def normalize_weights(weights):
    drift = weights.sum() / len(weights)
    return weights * (1 / drift)


def run_weighting_iteration(df, census='US', weigh_on=[], verbose=True):
    errors = []
    all_weights = {}

    if 'weight' not in df.columns:
        df.loc[:, 'weight'] = df[df.columns[0]].transform(lambda x: 1)
    
    census_data = get_census(census)

    if weigh_on == []:
        weigh_on = list(census_data.keys())

    for var in weigh_on:
        if var in df.columns:
            if var in census_data.keys():
                if verbose:
                    print('## {} ##'.format(var))
                    
                data = census_data[var]

                if df[var].isna().sum() > 0:
                    raise ValueError('{} contains NAs which are not allowed for weighting. Please impute or replace with an explicit NA value.'.format(var))

                counts = df[var].value_counts(normalize=True)
                missing_keys = list(set(counts.keys()) - set(data.keys()))
                if missing_keys:
                    data = pd.concat((pd.Series(data) * (1 - counts[missing_keys].sum()),
                                      counts[missing_keys]))
                    data = dict(data)

                prior_weight = df.groupby(var)['weight'].mean()
                weights = normalize_weights(pd.Series(data) / (counts * prior_weight))
                
                if verbose:
                    print(weights)
                    
                all_weights[var] = weights
                error = (df[var].replace(weights).apply(lambda x: np.abs(1 - x)).sum() / df.shape[0])
                errors.append(error)

                if verbose:
                    print('ERROR: {}'.format(error))
                    print('-')
                    print('-')
            elif verbose:
                print('-- WARNING: did not weigh on {} as it is not in census'.format(var))

        elif verbose:
            print('-- WARNING: did not weigh on {} as it is not in survey dataframe'.format(var))
            
            
    total_error = np.array(errors).sum()    
    error_table = sorted([(k,
                           np.array([np.abs(1 - v) for v in vs.values]).max()) for k, vs in all_weights.items()],
                         key=lambda x: x[1],
                         reverse=True)
    error_table = dict(error_table)
    if verbose:
        print(error_table)
        print('TOTAL ERROR: {}'.format(total_error))
        
    return {'errors': errors,
            'error_table': error_table,
            'weights': all_weights,
            'total_error': total_error}


def run_weighting_scheme(df, iters=10, census='US', weigh_on=[], early_terminate=True, verbose=1):
    df.loc[:, 'weight'] = df['age'].transform(lambda x: 1)
    iterx = 1
    weights = None
    last_var = None
    census_data = get_census(census)

    if weigh_on == []:
        weigh_on = list(census_data.keys())

    for var in weigh_on:
        if var in df.columns:
            if var in census_data.keys():
                output = run_weighting_iteration(df,
                                                 census=census,
                                                 weigh_on=weigh_on,
                                                 verbose=(verbose >= 2))

                total_error = output['total_error']

                if not weights and verbose >= 1:
                    print('ITER 1/{} - initialization - ERROR {}'.format(iterx, iters, var, total_error))

                weights = output['weights']
                df.loc[:, 'weight'] = df['weight'] * df[var].astype(str).replace(weights[var])
                df.loc[:, 'weight'] = normalize_weights(df['weight'])

                last_var = var
                iterx += 1
                if verbose >= 1:
                    print('ITER {}/{} - weight {} - ERROR {}'.format(iterx, iters, var, total_error))
            elif verbose >= 1:
                print('-- WARNING: did not weigh on {} as it is not in census'.format(var))
        elif verbose >= 1:
            print('-- WARNING: did not weigh on {} as it is not in survey dataframe'.format(var))
        if iterx > iters:
            break
    
    for i in range(iters - iterx):
        weigh_next = list(output['error_table'].keys())[0]
        if weigh_next == last_var and early_terminate:
            print('-- REACHED LOCAL MINIMUM; EARLY TERMINATION')
            break

        df.loc[:, 'weight'] = df['weight'] * df[weigh_next].astype(str).replace(weights[weigh_next])
        df.loc[:, 'weight'] = normalize_weights(df['weight'])

        output = run_weighting_iteration(df,
                                         census=census,
                                         weigh_on=weigh_on,
                                         verbose=(verbose >= 2))

        weights = output['weights']
        total_error = output['total_error']

        last_var = weigh_next
        iterx += 1

        if verbose >= 1:
            print('ITER {}/{} - weight {} - ERROR {}'.format(iterx,
                                                             iters,
                                                             weigh_next,
                                                             total_error))
        
    max_weight = df['weight'].max()
    min_weight = df['weight'].min()
    if verbose >= 1:
        print('Done - FINAL ERROR {} - MAX WEIGHT {} - MIN WEIGHT {}'.format(total_error,
                                                                             max_weight,
                                                                             min_weight))
    
    df.loc[:, 'weight'] = normalize_weights(df['weight'])

    return {'final_weights': df['weight'],
            'max_weight': max_weight,
            'min_weight': min_weight,
            'final_df': df,
            'error': total_error}



surveyweights/census/uk_census.py
UK_CENSUS = {'age': {'18-24': 0.094,
                     '25-29': 0.068,
                     '30-34': 0.066,
                     '35-39': 0.067,
                     '40-44': 0.073,
                     '45-49': 0.073,
                     '50-54': 0.064,
                     '55-59': 0.057,
                     '60-64': 0.060,
                     '65-69': 0.048,
                     '70-74': 0.039,
                     '75+': 0.078}, # https://www.ethnicity-facts-figures.service.gov.uk/uk-population-by-ethnicity/demographics/age-groups/latest

             'education': {'No qualifications': 0.23,
                           'Level 1 qualifications': 0.14,
                           'Level 2 qualifications': 0.15,
                           'Apprenticeship': 0.03,
                           'Level 3 qualifications': 0.12,
                           'Level 4 qualifications and above': 0.27,
                           'Other qualifications': 0.06}, # https://www.ons.gov.uk/employmentandlabourmarket/peopleinwork/employmentandemployeetypes/bulletins/keystatisticsandquickstatisticsforlocalauthoritiesintheunitedkingdom/2013-12-04

             'region': {'North East': 0.04,
                        'North West': 0.1098,
                        'Yorkshire And The Humber': 0.0824,
                        'East Midlands': 0.0725,
                        'West Midlands': 0.0889,
                        'East': 0.0935,
                        'London': 0.1342,
                        'South East': 0.1374,
                        'South West': 0.0844,
                        'Wales': 0.0472,
                        'Scotland': 0.0815,
                        'Northern Ireland': 0.0283},

             'gender': {'Female': 0.507,
                        'Male': 0.487,
                        'Other': 0.006}, # Male-Female from https://www.ethnicity-facts-figures.service.gov.uk/uk-population-by-ethnicity/demographics/male-and-female-populations/latest, other from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5227946/

             'income': {'Less than £15,000': 0.12,
					    '£15,000 to £20,000': 0.21,
					    '£20,000 and £30,000': 0.29,
                        '£30,000 to £50,000': 0.24,
                        '£50,000 to £70,000': 0.07,
                        '£70,000 to £100,000': 0.04,
                        'More than £100,000': 0.03},

             'vote_brexit': {'Remain': 0.4811,
                             'Leave': 0.5189}, # https://en.wikipedia.org/wiki/2016_United_Kingdom_European_Union_membership_referendum

             'vote2019': {'Conservative': 0.436,
                          'Labour': 0.321,
                          'SNP': 0.039, 
                          'Lib Dem': 0.116,
                          'Other': 0.088}} # 2019 UK election popular vote as recorded by Wikipedia https://en.wikipedia.org/wiki/2019_United_Kingdom_general_election


surveyweights/census/us_census.py
US_CENSUS = {'age': {'18-24': 0.1304,
                     '25-44': 0.3505,
                     '45-64': 0.3478,
                     '65+': 0.1713}, # Age from 2010 US Census https://www.census.gov/prod/cen2010/briefs/c2010br-03.pdf

             'education': {'Completed graduate school': 0.1204,
                           'Graduated from college': 0.2128,
                           'Some college, no degree': 0.2777,
                           'Graduated from high school': 0.2832,
                           'Less than high school': 0.1060}, # Education from 2019 ACS https://www.census.gov/data/tables/2019/demo/educational-attainment/cps-detailed-tables.html

             'gender': {'Female': 0.507,
                        'Male': 0.487,
                        'Other': 0.006}, # Male-Female from 2010 US Census https://www.census.gov/prod/cen2010/briefs/c2010br-03.pdf, other from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5227946/

             'income': {'Under $15,000': 0.1020,
                        'Between $15,000 and $49,999': 0.2970,
                        'Between $50,000 and $74,999': 0.1720,
                        'Between $75,000 and $99,999': 0.1250,
                        'Between $100,000 and $150,000': 0.1490,
                        'Over $150,000': 0.1550}, # Income from 2019 ACS (p34) https://www.census.gov/content/dam/Census/library/publications/2019/demo/p60-266.pdf

             'race': {'White or Caucasian': 0.6340,
                      'Asian or Asian American': 0.0590,
                      'Black or African American': 0.1340,
                      'Hispanic or Latino': 0.1530,
                      'Other': 0.02}, # Race from 2019 Census estimates https://en.wikipedia.org/wiki/Race_and_ethnicity_in_the_United_States#Racial_categories

             'urban_rural': {'Suburban': 0.5500,
                             'Urban': 0.3100,
                             'Rural': 0.1400}, # Urban-Rural from 2018 Pew https://www.pewsocialtrends.org/2018/05/22/demographic-and-economic-trends-in-urban-suburban-and-rural-communities/

			 # Regions follow a custom definition, see `US_REGIONS_MAP`
             'region': {'Midwest': 0.2149,
                        'Mountains': 0.0539,
                        'Northeast': 0.1604,
                        'Pacific': 0.1602,
                        'South': 0.2180,
                        'Southwest': 0.0531,
                        'Southeast': 0.1398}, # Regions from 2019 Census estimates https://en.wikipedia.org/wiki/List_of_states_and_territories_of_the_United_States_by_population

             'vote2016': {'Hillary Clinton': 0.482,
                          'Donald Trump': 0.461,
                          'Other': 0.057}, # 2016 US election popular vote as recorded by Wikipedia https://en.wikipedia.org/wiki/2016_United_States_presidential_election

             'vote2020': {'Joe Biden': 0.513,
                          'Donald Trump': 0.469,
                          'Other': 0.018}, # 2020 US election popular vote as recorded by Wikipedia https://en.wikipedia.org/wiki/2020_United_States_presidential_election

             'left_right': {'Liberal': 0.35,  # Sienna College / NYT poll <https://int.nyt.com/data/documenttools/nyt-siena-poll-methodology-june-2020/f6f533b4d07f4cbe/full.pdf>
                            'Moderate': 0.26,
                            'Conservative': 0.30},

             'gss_bible': {'Word of God': 0.31,
                           'Inspired word': 0.473,
                           'Book of fables': 0.217}, # From GSS 2018 https://gssdataexplorer.norc.org/variables/364/vshow

             'gss_trust': {'Can trust': 0.331,
                           'Can\'t be too careful': 0.669 }, # From GSS 2018 https://gssdataexplorer.norc.org/variables/441/vshow

             'gss_spanking': {'Agree': 0.677,
                              'Disagree': 0.323}} # From GSS 2018 https://gssdataexplorer.norc.org/trends/Gender%20&%20Marriage?measure=spanking


US_CA_CENSUS = {'age': {'18-24': 0.136,
                        '25-44': 0.434,
                        '45-64': 0.282,
                        '65+': 0.147}, # Age 2018 US Census https://www.infoplease.com/us/census/california/demographic-statistics

             'education': {'Completed graduate school': 0.13,
                           'Graduated from college': 0.218,
                           'Some college, no degree': 0.285,
                           'Graduated from high school': 0.206,
                           'Less than high school': 0.16}, # Education from https://www.statista.com/statistics/306963/educational-attainment-california

             'gender': {'Female': 0.499,
                        'Male': 0.495,
                        'Other': 0.006}, # Male-Female from 2018 US Census, other from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5227946/

             'income': {'Under $15,000': 0.106,
                        'Between $15,000 and $49,999': 0.2955,
                        'Between $50,000 and $74,999': 0.1652,
                        'Between $75,000 and $99,999': 0.1210,
                        'Between $100,000 and $150,000': 0.1523,
                        'Over $150,000': 0.1597}, # Income from https://statisticalatlas.com/state/California/Household-Income

             'race': {'White or Caucasian': 0.3664,
                      'Asian or Asian American': 0.1452,
                      'Black or African American': 0.0551,
                      'Hispanic or Latino': 0.3929,
                      'Other': 0.0404}, # Race from 2018 Census estimates https://en.wikipedia.org/wiki/File:Ethic_California_Organized_Pie.png

             'vote2016': {'Hillary Clinton': 0.6173,
                          'Donald Trump': 0.3162,
                          'Other': 0.0665}} # 2016 US election popular vote as recorded by Wikipedia https://en.wikipedia.org/wiki/2016_United_States_presidential_election_in_California

US_TX_CENSUS = {'age': {'18-24': 0.109,
                        '25-44': 0.389,
                        '45-64': 0.325,
                        '65+': 0.177}, # Age from 2019 US Census https://www.statista.com/statistics/912205/texas-population-share-age-group/
                                       # 72.7% over 18 per https://en.wikipedia.org/wiki/Demographics_of_Texas

             'education': {'Completed graduate school': 0.108,
                           'Graduated from college': 0.2,
                           'Some college, no degree': 0.287,
                           'Graduated from high school': 0.252,
                           'Less than high school': 0.154}, # Education from https://www.statista.com/statistics/306995/educational-attainment-texas/

             'gender': {'Female': 0.499,
                        'Male': 0.495,
                        'Other': 0.006}, # Male-Female from https://www.census.gov/quickfacts/TX, other from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5227946/

             'income': {'Under $15,000': 0.1186,
                        'Between $15,000 and $49,999': 0.3524,
                        'Between $50,000 and $74,999': 0.1652,
                        'Between $75,000 and $99,999': 0.119,
                        'Between $100,000 and $150,000': 0.1342,
                        'Over $150,000': 0.1106}, # Income from https://statisticalatlas.com/state/Texas/Household-Income

             'race': {'White or Caucasian': 0.4140,
                      'Asian or Asian American': 0.0493,
                      'Black or African American': 0.1188,
                      'Hispanic or Latino': 0.3961,
                      'Other': 0.0218}, # Race from 2018 Census estimates https://en.wikipedia.org/wiki/Demographics_of_Texas

             'vote2016': {'Hillary Clinton': 0.4324,
                          'Donald Trump': 0.5223,
                          'Other': 0.0453}} # 2016 US election popular vote as recorded by Wikipedia https://en.wikipedia.org/wiki/2016_United_States_presidential_election_in_Texas


US_REGIONS_MAP = {'Midwest': ['Illinois', 'Indiana', 'Iowa', 'Michigan', 'Minnesota',
                              'Ohio', 'Pennsylvania', 'Wisconsin'],
                  'Mountains': ['Alaska', 'Idaho', 'Kansas', 'Montana', 'Nebraska',
                                'North Dakota', 'South Dakota', 'Utah', 'Wyoming', 'Oklahoma'],
                  'Northeast': ['Connecticut', 'Delaware', 'District of Columbia (DC)',
                                'Maine', 'Maryland', 'Massachusetts', 'New Hampshire',
                                'New Jersey', 'New York', 'Rhode Island', 'Vermont'],
                  'Pacific': ['California', 'Hawaii', 'Oregon', 'Washington', 'Guam',
                              'Puerto Rico', 'Virgin Islands'],
                  'South': ['Missouri', 'Tennessee', 'Alabama', 'Arkansas', 'Kentucky',
                            'Louisiana', 'Mississippi', 'Texas', 'Virginia', 'West Virginia'],
                  'Southwest': ['Arizona', 'Colorado', 'Nevada', 'New Mexico'],
                  'Southeast': ['Florida', 'Georgia', 'North Carolina', 'South Carolina']}

